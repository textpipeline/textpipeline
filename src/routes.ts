import { matchPath, RouteProps } from 'react-router-dom';

interface ExtendedRouteProps extends RouteProps {
  readonly linkPath: string;
}

const routes: Record<string, ExtendedRouteProps> = {
  home: {
    path: '/',
    linkPath: '/',
    exact: true,
  },
};

export const isKnownPath = (suspectPath: string): boolean =>
  Object.keys(routes)
    .map(key => routes[key])
    .some(route => matchPath(suspectPath, route));

export default routes;
