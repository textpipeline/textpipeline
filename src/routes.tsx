import React from 'react';
import { matchPath, RouteProps } from 'react-router-dom';

import Home from './components/Home';
import OpenSource from './components/OpenSource';
import OpenSourceProject from './components/OpenSourceProject';
import Transform from './components/Transform';
import { bySlug as ossBySlug } from './oss';
import { bySlug as transformsBySlug } from './transforms';

type BaseRouteKey = 'home' | 'oss';

type PathableRouteProps = RouteProps & { path: string };

export const homePath = '/';
export const ossPath = '/oss';
export const transformsPath = '/transforms';

const baseRoutes: Record<BaseRouteKey, PathableRouteProps> = {
  home: {
    path: homePath,
    component: Home,
    exact: true,
  },
  oss: {
    path: ossPath,
    component: () => <OpenSource ossPath={ossPath} />,
    exact: true,
  },
};

const transformRoutes: Record<string, PathableRouteProps> = Object.keys(transformsBySlug).reduce(
  (transformRoutesSoFar, slug) => {
    const { name, inputTypeName, outputTypeName, openSourceProject, defaultOutput, execute } = transformsBySlug[slug];
    return {
      ...transformRoutesSoFar,
      [`transform-${slug}`]: {
        path: `${transformsPath}/${slug}`,
        component: () => (
          <Transform
            name={name}
            ossHref={openSourceProject ? `${ossPath}/${openSourceProject.slug}` : undefined}
            inputTypeName={inputTypeName}
            outputTypeName={outputTypeName}
            defaultOutput={defaultOutput}
            execute={execute}
          />
        ),
        exact: true,
      },
    };
  },
  {}
);

const ossRoutes: Record<string, PathableRouteProps> = Object.keys(ossBySlug).reduce((ossRoutesSoFar, slug) => {
  const { name, description, href, repositoryHref, license, licenseText } = ossBySlug[slug];
  return {
    ...ossRoutesSoFar,
    [`oss-${slug}`]: {
      path: `${ossPath}/${slug}`,
      component: () => (
        <OpenSourceProject
          name={name}
          description={description}
          slug={slug}
          href={href}
          repositoryHref={repositoryHref}
          license={license}
          licenseText={licenseText}
          transformsPath={transformsPath}
        />
      ),
      exact: true,
    },
  };
}, {});

const routes: Record<string, PathableRouteProps> = {
  ...baseRoutes,
  ...transformRoutes,
  ...ossRoutes,
};

export const isKnownPath = (suspectPath: string): boolean =>
  Object.keys(routes)
    .map(key => routes[key])
    .some(route => matchPath(suspectPath, route));

export default routes;
