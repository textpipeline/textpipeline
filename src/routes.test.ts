import routes, { isKnownPath } from './routes';

describe('routes', () => {
  describe('routes', () => {
    it('has the correct routes', () => {
      expect(routes).toMatchSnapshot();
    });
  });

  describe('isKnownPath', () => {
    it('returns true for a known path', () => {
      expect(isKnownPath('/')).toBe(true);
    });

    it('returns false for an unknown path', () => {
      expect(isKnownPath('/barglesnarf')).toBe(false);
    });
  });
});
