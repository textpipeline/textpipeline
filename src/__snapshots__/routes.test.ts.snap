// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`routes routes has the correct routes 1`] = `
Object {
  "home": Object {
    "exact": true,
    "linkPath": "/",
    "path": "/",
  },
}
`;
